javalib/src/main/scala/java/util/TreeSet.scala

```diff
@@ -0,0 +1,219 @@
+package java.util
+
+import java.lang.Comparable
+import scala.collection.mutable
+import scala.math.Ordering
+
+import scala.collection.JavaConversions._
+
+class TreeSet[E] (_comparator: Comparator[_ >: E])
+    extends AbstractSet[E]
+    with NavigableSet[E]
+    with Cloneable
+    with Serializable { self =>
+
+  def this() =
+    this(null.asInstanceOf[Comparator[_ >: E]])
+
+  def this(collection: Collection[_ <: E]) = {
+    this(null.asInstanceOf[Comparator[E]])
+    addAll(collection)
+  }
+
+  def this(sortedSet: SortedSet[E]) = {
+    this(sortedSet.comparator())
+    addAll(sortedSet)
+  }
+
+  private implicit object BoxOrdering extends Ordering[Box[E]] {
+    def compare(a: Box[E], b:Box[E]): Int = {
+      if (_comparator ne null) _comparator.compare(a.inner, b.inner)
+      else defaultOrdering[E].compare(a.inner, b.inner)
```

Comment:
Here you are creating a new `defaultOrdering` instance each time you compare this might be a performance killer. Additionally you are checking on each test `if (_comparator ne null)` which could be done once on the creation of the object. You could solve this with: 

```scala
private implicit object BoxOrdering extends Ordering[Box[E]] {
  val cmp = {
    if (_comparator ne null) _comparator
    else defaultOrdering[E]
  }
  def compare(a: Box[E], b:Box[E]): Int = cmp.compare(a.inner, b.inner)
}
```